runtime/gc_structs/ClassLoaderClassesIterator.cpp

```diff
@@ -108,9 +123,13 @@ GC_ClassLoaderClassesIterator::nextClass()
 		if (ANONYMOUS_CLASSES == _mode) {
 			_nextClass = nextAnonymousClass();
 		} else {
-			if ( (result->classLoader == _classLoader) && (NULL != result->arrayClass) ) {
+			if ((result->classLoader == _classLoader) && (NULL != result->arrayClass)) {
 				/* this class is defined in the loader, so follow its array classes */
 				_nextClass = result->arrayClass;
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+			} else if ((result->classLoader == _classLoader) && (NULL != getNullRestrictedArray(result))) {
```

Comment:
Spoke with Hang. Given that the higher order J9Class arrays are different (ie. Foo[][] != Foo[][!], etc.) then for a given J9Class we need to potentially iterate two lists. The normal array list and the NR list

One approach is to add a new state `VALUE_CLASSES` for the Null Restricted case where, clazz->NRArray is walked (and then the subsequent arrayClazz->arrayClazz) until NULL is reached. At this point, we will return to the base class via leafComponentType then switch back to `TABLE_CLASSES` mode. In this mode, it should work like it currently does.

Here is an example:

```
	if (NULL != result) {
		if (ANONYMOUS_CLASSES == _mode) {
			_nextClass = nextAnonymousClass();
		} else {
			if ((result->classLoader == _classLoader) && (NULL != result->nullRestrictedArrayClass)) {
				_mode = VALUE;
				_nextClass = result->nullRestrictedArrayClass
			} else if ( (result->classLoader == _classLoader) && (NULL != result->arrayClass) ) {
				/* this class is defined in the loader, so follow its array classes */
				_nextClass = result->arrayClass;
			} else if (VALUE == _mode) {
				J9ArrayClass *arrayClazz = result->leafComponentType->arrayClass;
				_mode = TABLE_CLASSES;
				if ( (result->classLoader == _classLoader) && (NULL != arrayClazz)) {
					_nextClass = arrayClazz;
				} else {
					_nextClass = nextTableClass();
				}
			} else if (TABLE_CLASSES == _mode) {
				_nextClass = nextTableClass();
			} else {
				_nextClass = nextSystemClass();
			}
		}
	}
```

Another option is we create an `ARRAY_MODE` and encapsulate all array walking logic inside a `nextArrayClass()` function. This would clean up the code above, but we would still require a `VALUE` mode or an equivalent substate to walk the two lists.

@dmitripivkine do you have a preference on the suggested approach?

Reply:
I am not sure use `_mode` is a good idea - you need to preserve original value (ex. `TABLE_CLASSES`) and restore it after. I think (maybe) creating an independent `_state` will serve better. For example `_mode` is going to stay unchanged (and there is no assumption only `TABLE_CLASSES` might have Null Restricted). You can switch between "Array" and "VT" states and still be in the current mode. Will it work?

Reply:
>  creating an independent _state will serve better.

Yes this will work.

Reply:
I updated my changes to include this. Please take a look.

Reply:
> I updated my changes to include this. Please take a look.

I think what Dmitri means by _state are enum like https://github.com/eclipse-openj9/openj9/blob/1cac5050a7914a28fbab9b1e8e1aea9c16eeaa5a/runtime/gc_structs/ClassArrayClassSlotIterator.hpp#L47-L52

Use "Array" state when walking the normal arrayclass list, use "VT" or "NullrestirctiedArray" state when walking the nullrestricted array list.

Reply:
I figured since there's only two states a boolean would be sufficient to describe that.